### PR TITLE
Lvruby 28

### PR DESCRIPTION
**Pull request by AbdulRustam**
## JIRA ticket

[LVRUBY-28](http://ssu-jira.softserveinc.com/secure/RapidBoard.jspa?rapidView=979&view=planning&selectedIssue=LVRUBY-28)

## Code reviewers

_Please list your code reviewers here._

- [ ] @AlinaKovtun  
- [ ] @omaik 

## Summary of issue

As user, I want to see my profile that other users can't see
As user, I want to change my data

## Summary of change

I create profile page, the user can view, change their data and view your news

## Review checklist

- [x] Rubocop doesn't raise any warnings and offenses
- [ ] New features are covered with new specs
- [ ] All specs pass

### CODE REVIEW

**Comment by vikabelous** [comment]
**Code Coverage:**
:red_circle: **75.86%** covered at **0.75** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:red_circle: **75.86%** covered at **0.75** hits/line

File: app/views/user_mailer/change_email_password_confirmation.html.haml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+%h1= @user.email
+
+= link_to "Profile", @user
```
Comment (AlinaKovtun) [inline app/views/user_mailer/change_email_password_confirmation.html.haml]: add newline

File: spec/controllers/users_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+require 'rails_helper'
+require 'factory_bot'
+
+RSpec.describe UsersController, type: :controller do
+  let(:user) { build(:user) }
+  let(:password) { 'wordpass1' }
+
+  it 'change password' do
+    # user.update(password: 'wordpass1')
```
Comment (AlinaKovtun) [inline spec/controllers/users_controller_spec.rb]: remove unnecessary comments

File: spec/helpers/users_helper_spec.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+# Specs in this file have access to a helper object that includes
```
Comment (AlinaKovtun) [inline spec/helpers/users_helper_spec.rb]: remove unnecessary comments

**Comment by vikabelous** [comment]
**Code Coverage:**
:red_circle: **75.0%** covered at **0.74** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:red_circle: **75.0%** covered at **0.74** hits/line

File: .circleci/config.yml [hunk 4]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (omaik) [inline .circleci/config.yml]: why do you need changes in this file?

File: app/assets/javascripts/users.coffee [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (omaik) [inline app/assets/javascripts/users.coffee]: please remove files which are not used

File: app/assets/stylesheets/application.css [hunk 6]
```diff
@@ -13,3 +13,14 @@
  *= require_tree .
  *= require_self
  */
+
+@import "material";
+@import "https://fonts.googleapis.com/icon?family=Material+Icons";
+
+input {
```
Comment (omaik) [inline app/assets/stylesheets/application.css]: as was mentioned for Oleksandr, please move styles to the file which respond to the page they are used in.

Also, the current style will affect all input elements on all pages, I assume this is not good.

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+class UsersController < ApplicationController
+  def profile; end
+
+  def edit; end
+
+  def show; end
```
Comment (omaik) [inline app/controllers/users_controller.rb]: what is the difference between `profile` and `show`

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+  def update
+    if current_user.update(user_params)
+      redirect_to user_path(current_user.id)
+    else
+      render 'edit'
```
Comment (omaik) [inline app/controllers/users_controller.rb]: in this case, you should also render some errors, to help the user identify what he did wrong

File: config/application.rb [hunk 9]
```diff
@@ -19,3 +19,13 @@ class Application < Rails::Application
     # the framework and any gems in your application.
   end
 end
+
+ActionMailer::Base.smtp_settings = {
```
Comment (omaik) [inline config/application.rb]: do you need this both here and in `development.rb`?

File: config/routes.rb [hunk 10]
```diff
@@ -3,5 +3,9 @@
 Rails.application.routes.draw do
   devise_for :users
 
-  root to: 'application#welcome'
+  resources :users
+  get 'profile', to: 'users#profile'
+  get 'edit_profile', to: 'users#edit'
```
Comment (omaik) [inline config/routes.rb]: this route does not follow REST principles. And also, please use a symbol instead of string

File: config/routes.rb [hunk 11]
```diff
@@ -3,5 +3,9 @@
...
+
+  root to: 'users#profile'
```
Comment (omaik) [inline config/routes.rb]: And how this will be handled for not logged in users?

File: spec/controllers/users_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+require 'rails_helper'
+require 'factory_bot'
+
+RSpec.describe UsersController, type: :controller do
+  let(:user) { build(:user) }
+  let(:pass) { 'wordpass1' }
+
+  it 'change password' do
+    user.skip_confirmation_notification!
```
Comment (omaik) [inline spec/controllers/users_controller_spec.rb]: in this test, you do not invoke controller code at all, so what do you test?)

File: spec/controllers/users_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+  it 'not change password' do
+    user.skip_confirmation_notification!
+    user.update(password: '')
+    expect(user.valid_password?(user.password)).to be(false)
+  end
+
+  describe '#update' do
+
+    context 'PUT #update' do
+      let!(:user1) { build(:user) }
```
Comment (omaik) [inline spec/controllers/users_controller_spec.rb]: `let!` seems redundant

File: spec/controllers/users_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+      it 'should update user password' do
+        params = {
+          email: user1.email,
+          password: pass,
+          password_confirmation: pass
+        }
+        user1.skip_confirmation_notification!
+        put :update, params: { params }
+        params.keys.each do |key|
+          expect(user.attributes[key.to_s]).to eq params[key]
```
Comment (omaik) [inline spec/controllers/users_controller_spec.rb]: seeing this test I have a question. How the password is stored in DB, do you hash it?

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **100.0%** covered at **1.12** hits/line

File: spec/controllers/users_controller_spec.rb (hunk 14 above)
Comment (AbdulRustam) [reply spec/controllers/users_controller_spec.rb]: I refactor test

File: config/routes.rb (hunk 11 above)
Comment (AbdulRustam) [reply config/routes.rb]: I delete routes

File: app/controllers/users_controller.rb (hunk 8 above)
Comment (AbdulRustam) [reply app/controllers/users_controller.rb]: In view

File: app/controllers/users_controller.rb (hunk 7 above)
Comment (AbdulRustam) [reply app/controllers/users_controller.rb]: I delete profile

File: app/assets/stylesheets/application.css (hunk 6 above)
Comment (AbdulRustam) [reply app/assets/stylesheets/application.css]: I create new file for users

File: app/assets/javascripts/users.coffee (hunk 5 above)
Comment (AbdulRustam) [reply app/assets/javascripts/users.coffee]: Ok

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **93.33%** covered at **1.8** hits/line

File: config/application.rb [hunk 15]
```diff
@@ -15,11 +15,21 @@ class Application < Rails::Application
 
     # Settings in config/environments/* take precedence over those specified here.
     # Application configuration can go into files in config/initializers
-    # -- all .rb files in that directory are automatically loaded after loading
+# -- all .rb files in that directory are automatically loaded after loading
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply config/application.rb]: add tab

File: config/application.rb (hunk 9 above)
Comment (AbdulRustam) [reply config/application.rb]: No, I delete this code

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    end
+  end
+
+  private
+
+  def password_update
+    unless password_params[:password].blank?
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/controllers/users_controller.rb]: I refactoring code

File: config/routes.rb (hunk 10 above)
Comment (AbdulRustam) [reply config/routes.rb]: I delete this routes

File: spec/controllers/users_controller_spec.rb (hunk 12 above)
Comment (AbdulRustam) [reply spec/controllers/users_controller_spec.rb]: I write new test

File: spec/controllers/users_controller_spec.rb (hunk 13 above)
Comment (AbdulRustam) [reply spec/controllers/users_controller_spec.rb]: I delete this let!

File: app/views/users/show.html.haml [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+          %hr/
+          .row
+            .col-sm-3.col-md-2.col-5
+              %label{style: 'font-weight:bold;'} Last Name
+            .col-md-8.col-6
+              = current_user.last_name
+          %hr/
+          .row
+            .col-sm-3.col-md-2.col-5
+              %label{style: 'font-weight:bold;'} About me
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/show.html.haml]: I delete this style

File: app/views/users/show.html.haml [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+          %hr/
+          .row
+            .col-sm-3.col-md-2.col-5
+              %label{style: 'font-weight:bold;'} First Name
+            .col-md-8.col-6
+              = current_user.first_name
+          %hr/
+          .row
+            .col-sm-3.col-md-2.col-5
+              %label{style: 'font-weight:bold;'} Last Name
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/show.html.haml]: I delete this style

File: app/views/users/show.html.haml [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+          %hr
+          .row
+            .col-sm-3.col-md-2.col-5
+              %label{style: 'font-weight:bold;'} Email
+            .col-md-8.col-6
+              = current_user.email
+          %hr/
+          .row
+            .col-sm-3.col-md-2.col-5
+              %label{style: 'font-weight:bold;'} First Name
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/show.html.haml]: I delete this style

File: app/views/users/edit.html.haml [hunk 20]
```diff
@@ -0,0 +1,47 @@
+= form_for current_user do |form|
+  .container
+    .row
+      .col-12
+        .card
+          #myTabContent.tab-content.ml-1
+            %hr
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} Email
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/edit.html.haml]: I delete this style

File: app/views/users/show.html.haml [hunk 21]
```diff
@@ -0,0 +1,34 @@
+.container
+  .row
+    .col-12
+      .card
+        #myTabContent.tab-content.ml-1
+          %hr
+          .row
+            .col-sm-3.col-md-2.col-5
+              %label{style: 'font-weight:bold;'} Email
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/show.html.haml]: I delete this style

File: app/views/users/edit.html.haml [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+            %hr/
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} Last Name
+              .col-md-8.col-6
+                = form.text_field :last_name, class: 'form-control'
+            %hr/
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} About me
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/edit.html.haml]: I delete this style

File: app/views/users/edit.html.haml [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+            %hr/
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} First Name
+              .col-md-8.col-6
+                = form.text_field :first_name, class: 'form-control'
+            %hr/
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} Last Name
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/edit.html.haml]: I delete this style

File: app/views/users/edit.html.haml [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+            %hr/
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} Password Confirmation
+              .col-md-8.col-6
+                = form.password_field :password_confirmation, class: 'form-control'
+            %hr/
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} First Name
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/edit.html.haml]: I delete this style

File: app/views/users/edit.html.haml [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+            %hr/
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} Password
+              .col-md-8.col-6
+                = form.password_field :password, class: 'form-control'
+            %hr/
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} Password Confirmation
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/edit.html.haml]: I delete this style

File: app/views/users/edit.html.haml [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+            %hr
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} Email
+              .col-md-8.col-6
+                = form.text_field :email, class: 'form-control' , :readonly => true
+            %hr/
+            .row
+              .col-sm-3.col-md-2.col-5
+                %label{style: 'font-weight:bold;'} Password
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/edit.html.haml]: I delete this style

File: app/views/users/edit.html.haml [hunk 27]
```diff
@@ -0,0 +1,47 @@
+= form_for current_user do |form|
+  .container
+    .row
+      .col-12
+        .card
+          #myTabContent.tab-content.ml-1
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/views/users/edit.html.haml]: i delete #myTabContext

File: app/controllers/users_controller.rb [hunk 28]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  private
+
+  def password_params
+    params.require(:user).permit(:password, :password_confirmation)
+  end
+
+  def user_params
+    params.require(:user).permit( :last_name, :first_name, :about_me)
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/controllers/users_controller.rb]: I delete this space

File: app/controllers/users_controller.rb [hunk 29]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class UsersController < ApplicationController
+  def edit; end
+
+  def show; end
+
+  def update
+    unless password_params[:password].blank?
```
[the remark this answers is not in the document]
Comment (AbdulRustam) [reply app/controllers/users_controller.rb]: I refactor this code

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **93.33%** covered at **1.8** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.87%** covered at **1.96** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.87%** covered at **1.96** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.87%** covered at **1.96** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **94.87%** covered at **1.96** hits/line

**Pull request closed without merging**